### PR TITLE
REF: sql insert_data operate column-wise to avoid internals

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/io/sql.py
Comment (WillAyd) [inline pandas/io/sql.py:696]: Is the issue with using `.items` that it can provide a frame back in case of duplicate labels? If so wonder if we shouldn't just update `.items` to do what you have done here

Comment (jbrockmendel) [reply pandas/io/sql.py:696]: you mean BlockManager.items?  i dont see how thats relevant here

Comment (WillAyd) [reply pandas/io/sql.py:696]: `temp` is a DataFrame, no?

Comment (jbrockmendel) [reply pandas/io/sql.py:696]: yes.  this is just doing this column-wise instead of block-wise to avoid accessing `._data`

File: pandas/io/sql.py
Comment (jbrockmendel) [reply pandas/io/sql.py:698]: sure, _ixs works here

Comment (jorisvandenbossche) [reply pandas/io/sql.py:698]: Although _ixs helps, the big win actually comes from avoiding creating a series if all you need are the values (as I did in the linked snippet)

Comment (jorisvandenbossche) [reply pandas/io/sql.py:698]: https://github.com/pandas-dev/pandas/pull/33252

Comment (jbrockmendel) [reply pandas/io/sql.py:698]: OK.  Until #33252 goes through, any objection to _ixs here?

Comment (jorisvandenbossche) [reply pandas/io/sql.py:698]: Yes, whichever gets merged second can be update to use it here

Comment (jreback) [reply pandas/io/sql.py:698]: I really don't like using private methods like this, what is wrong with .iloc here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/sql.py
Comment (jreback) [reply pandas/io/sql.py:698]: Using methods *internally* in frame.py is one thing; however this should be using a public API (and likely this makes almost no difference here).

Comment (jbrockmendel) [reply pandas/io/sql.py:698]: i agree with jeff on this one

Comment (jorisvandenbossche) [reply pandas/io/sql.py:698]: It's exactly the goal of the helper function I am adding in https://github.com/pandas-dev/pandas/pull/33252 to be used in situations where only the array values are needed, like here is the case.

So if we are adding such helper method, why not use it? (the reason I did the PR is because I want to see it used in several places, and many of those places will be *outside* of frame.py, eg in the ops code)

If the privateness of the method is a problem, let's discuss making it public. Although I personally think that is not needed right now (and we are using plenty of semi-private methods on DataFrame/Series outside of `frame.py` already)

Comment (jbrockmendel) [reply pandas/io/sql.py:698]: lets revisit that once #33252 is merged

Comment (jorisvandenbossche) [reply pandas/io/sql.py:698]: This PR certainly does not need to wait on #33252 being merged, but I still first want to have the discussion whether we find this an appropriate place to use that function here, once it exists.

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

File: pandas/io/sql.py
Comment (jreback) [reply pandas/io/sql.py:698]: I don't think we should be using internal functions *except* in the internals. This is not internal by any stretch (would say that pandas.io is offlimits entirely)

File: pandas/io/sql.py
Comment (jreback) [inline pandas/io/sql.py:697]: you should just use .items() here no? which yields the name and column

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/sql.py
Comment (jbrockmendel) [reply pandas/io/sql.py:697]: we're going to need `i` below

Comment (WillAyd) [reply pandas/io/sql.py:697]: Hinted at this in my first comment but I do also agree it would be nice to use `items` - we seem to special case it a lot (I've done so in groupby) I think just to work around the fact that it can return a DataFrame for duplicate labels

If we didn't do that maybe we could just enumerate over `df.items()` and have one way of doing things

Comment (jbrockmendel) [reply pandas/io/sql.py:697]: > Hinted at this in my first comment

Oh! you meant DataFrame.items!  I thought you were referring to BlockManager.items, which matches DataFrame.columns

Comment (jreback) [reply pandas/io/sql.py:697]: yes (with an enumerate on top if you need the i)

Comment (jbrockmendel) [reply pandas/io/sql.py]: updated to use enumerate and .items

**Comment by jorisvandenbossche** [comment]
> I don't think we should be using internal functions except in the internals. This is not internal by any stretch (would say that pandas.io is offlimits entirely)

Can you please explain a bit more *why* you don't want a method like `DataFrame._iter_column_arrays` to be used outside of `frame.py` ?

Yes, we want to use as much public functions in the other pandas modules. But we do have a set of "private" methods that we use ourselves, that I would see as "internal developer APIs". For example, we also use `Series._values`, `Series._can_hold_na`, `Index._get_level_values` (this are a few examples currently being used in sql.py), and many others, outside core/series.py or core/indexes/. 
Or do you find that those should also not be used? 

Or, do you want to limit the use here because it is outside pandas/core? (initially, you commented you didn't want private DataFrame methods to be used outside of frame.py, but maybe you meant outside pandas/core, and within all of /core it is fine to use those? In which case the argument would only be about pandas.io, pandas.plotting, pandas.tseries, more or less) 
Just trying to understand your argument.

**Comment by jbrockmendel** [comment]
> Can you please explain a bit more why you don't want a method like DataFrame._iter_column_arrays to be used outside of frame.py ?

Can we please bike-shed this elsewhere?

File: pandas/io/sql.py
Comment (WillAyd) [reply pandas/io/sql.py]: This looks good. Only comment I have is that this probably will fail with duplicate column labels if not already tested.

Comment (jbrockmendel) [reply pandas/io/sql.py]: DataFrame.items has a check for columns.is_unique; is that check not enough?

Comment (WillAyd) [reply pandas/io/sql.py]: Ah thanks; misunderstanding on my end

**Review by WillAyd — APPROVED** [review]
lgtm I think a nice cleanup

**Pull request merged**
